File: posix/samples/arrayloops.f
C-----------------------------------------------------------------------------
C FILE: arrayloops.f
C DESCRIPTION:
C   Example code demonstrating decomposition of array processing by
C   distributing loop iterations.  A global sum is maintained by a mutex
C   variable.  The Fortran example can be compiled with the command (IBM only):
C   xlf_r -O2 -qsmp=omp -qnosave arrayloops.f
C   Note also the need for the f_pthread module.
C AUTHOR: Blaise Barney
C LAST REVISED: 02/11/2002 Blaise Barney
C-----------------------------------------------------------------------------

      program arrayloops

C     This IBM Pthreads library module is required
      use f_pthread

C     Globals
      integer NTHREADS, ARRAYSIZE, ITERATIONS
      parameter(NTHREADS=4)
      parameter(ARRAYSIZE=1000000)
      parameter(ITERATIONS=ARRAYSIZE/NTHREADS)
      real*8 A(ARRAYSIZE), sum
      type(f_pthread_mutex_t) sum_mutex
      common /global/ A, sum, sum_mutex

C     Local variables
      integer rc, i, tids(NTHREADS)
      type(f_pthread_t) threads(NTHREADS)
      type(f_pthread_attr_t) attr

C     This will be the routine which threads execute
      external do_work

C     Pthreads setup:  Initialize mutex and explicitly create threads in a
C     joinable state (for portability).  Pass each thread its loop  offset
      rc = f_pthread_mutex_init(sum_mutex)
      rc = f_pthread_attr_init(attr)
      rc = f_pthread_attr_setdetachstate(attr,
     +                               PTHREAD_CREATE_JOINABLE)
      sum = 0
      do 100 i=1, NTHREADS
      tids(i) = i
      rc = f_pthread_create(threads(i),attr,FLAG_DEFAULT,
     +                      do_work,tids(i))
  100 continue

C     Wait for all threads to complete then print global sum 
      do 200 i=1,NTHREADS
        rc = f_pthread_join(threads(i))
  200 continue
      write(*,*)'Done. Sum= ',sum

C     Clean up and exit
      rc = f_pthread_attr_destroy(attr)
      rc = f_pthread_mutex_destroy(sum_mutex)
      call f_pthread_exit(rc)
      end

C-----------------------------------------------------------------------------

      subroutine do_work(mytid)

C     This IBM Pthreads library module is required
      use f_pthread

C     Globals
      integer NTHREADS, ARRAYSIZE, ITERATIONS
      parameter(NTHREADS=4)
      parameter(ARRAYSIZE=1000000)
      parameter(ITERATIONS=ARRAYSIZE/NTHREADS)
      real*8 A(ARRAYSIZE), sum
      type(f_pthread_mutex_t) sum_mutex
      common /global/ A, sum, sum_mutex

C     Local variables
      integer rc, mytid, start, end, i
      real*8 mysum

C     Initialize my part of the global array and keep local sum 
      start = (mytid-1) * ITERATIONS +1
      end = start + ITERATIONS - 1
      write(*,*) 'Thread ',mytid,' doing iterations ',start,
     +           ' to ',end
      mysum = 0
      do 100 i=start,end
        A(i) = i * 1.0
        mysum = mysum + a(i)
  100 continue

C     Lock the mutex and update the global sum, then exit
      rc = f_pthread_mutex_lock(sum_mutex)
      sum = sum + mysum
      rc = f_pthread_mutex_unlock(sum_mutex)
      call f_pthread_exit(rc)
      end subroutine do_work

File: posix/samples/arrayloops.c
/******************************************************************************
* FILE: arrayloops.c
* DESCRIPTION:
*   Example code demonstrating decomposition of array processing by
*   distributing loop iterations.  A global sum is maintained by a mutex
*   variable.  
* AUTHOR: Blaise Barney
* LAST REVISED: 01/29/09
******************************************************************************/
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>

#define NTHREADS      4
#define ARRAYSIZE   1000000
#define ITERATIONS   ARRAYSIZE / NTHREADS

double  sum=0.0, a[ARRAYSIZE];
pthread_mutex_t sum_mutex;


void *do_work(void *tid) 
{
  int i, start, *mytid, end;
  double mysum=0.0;

  /* Initialize my part of the global array and keep local sum */
  mytid = (int *) tid;
  start = (*mytid * ITERATIONS);
  end = start + ITERATIONS;
  printf ("Thread %d doing iterations %d to %d\n",*mytid,start,end-1); 
  for (i=start; i < end ; i++) {
    a[i] = i * 1.0;
    mysum = mysum + a[i];
    }

  /* Lock the mutex and update the global sum, then exit */
  pthread_mutex_lock (&sum_mutex);
  sum = sum + mysum;
  pthread_mutex_unlock (&sum_mutex);
  pthread_exit(NULL);
}


int main(int argc, char *argv[])
{
  int i, start, tids[NTHREADS];
  pthread_t threads[NTHREADS];
  pthread_attr_t attr;

  /* Pthreads setup: initialize mutex and explicitly create threads in a
     joinable state (for portability).  Pass each thread its loop offset */
  pthread_mutex_init(&sum_mutex, NULL);
  pthread_attr_init(&attr);
  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
  for (i=0; i<NTHREADS; i++) {
    tids[i] = i;
    pthread_create(&threads[i], &attr, do_work, (void *) &tids[i]);
    }

  /* Wait for all threads to complete then print global sum */ 
  for (i=0; i<NTHREADS; i++) {
    pthread_join(threads[i], NULL);
  }
  printf ("Done. Sum= %e \n", sum);

  sum=0.0;
  for (i=0;i<ARRAYSIZE;i++){ 
  a[i] = i*1.0;
  sum = sum + a[i]; }
  printf("Check Sum= %e\n",sum);

  /* Clean up and exit */
  pthread_attr_destroy(&attr);
  pthread_mutex_destroy(&sum_mutex);
  pthread_exit (NULL);
}

